### PR TITLE
Telemetry module implementation

### PR DESCRIPTION
**Pull request by vpereverzev**
- added new module "telemetry"
- added new module "global" for utility purpose with general types, interfaces or QML components
- added a dependency injection implementation
- added a modules setup semantic

Detailed explanation for community will be added to musescore/dev-handbook later

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
I believe all those 'merge' commits should not be in this PR.

File: mscore/shortcut.cpp [hunk 1]
```diff
@@ -4359,7 +4360,25 @@ void Shortcut::resetToDefault()
...
+        QAction* action = shortcut->action();
+
+        if (!action) {
+            continue;
+        }
+
+        if (action->shortcut() == keySequence) {
+            return shortcut;
+        }
+    }
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: Banner style indent of closing braces please. In this case many of those braces are not needed al all, as they bracket a singlc instrunction only

File: mscore/shortcut.cpp [hunk 2]
```diff
@@ -14,6 +14,7 @@
 #include "shortcut.h"
 #include "icons.h"
 #include "libmscore/xml.h"
+#include <QHash>
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.cpp]: Isn't that something for all.h?

File: mscore/shortcut.h [hunk 3]
```diff
@@ -63,6 +63,8 @@ Shortcuts marked with the STATE_NEVER state should NEVER used directly as shortc
 
 #include "icons.h"
 #include "globals.h"
+#include <QKeySequence>
+#include <QAction>
```
Comment (Jojo-Schmitz) [inline mscore/shortcut.h]: Isn't this something for all.h?

File: global/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+get_filename_component(DIR_NAME ${CMAKE_CURRENT_SOURCE_DIR} NAME)
+message("Configuring " ${DIR_NAME})
+
+find_package(Qt5 COMPONENTS Core REQUIRED CONFIG)
+add_definitions(
+    ${Qt5Core_DEFINITIONS}
+)
+include_directories(
+    ${Qt5Core_INCLUDE_DIRS}
+)
```
Comment (Jojo-Schmitz) [inline global/CMakeLists.txt]: I believe here we use banner style indent too

File: global/interfaces/abstractmodulesetup.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+#ifndef IMODULESETUP_H
+#define IMODULESETUP_H
+
+#include <QString>
+#include <QtQml/qqml.h>
+
+class AbstractModuleSetup
+      {
+public:
+      void setup(){
```
Comment (Jojo-Schmitz) [inline global/interfaces/abstractmodulesetup.h]: space between ) and {

File: global/interfaces/abstractmodulesetup.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            }
+
+protected:
+      virtual QString moduleName() = 0;
+
+      virtual void registerExports(){}
+
+      virtual void registerResources(){}
+
+      virtual void registerQmlTypes(){}
```
Comment (Jojo-Schmitz) [inline global/interfaces/abstractmodulesetup.h]: the above 3 too

File: global/servicesresolver.h [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+#include <QHash>
+#include <QMetaType>
+#include <QUuid>
+
+#define INTERFACE_ID                               \
+      public:                                      \
+      static const QUuid interfaceId() {           \
+          static QUuid id = QUuid::createUuid();   \
+          return id;                               \
+      }                                            \
```
Comment (Jojo-Schmitz) [inline global/servicesresolver.h]: Mabber stakly indend, here and on other places. And 6 spaces, not 4

File: global/serviceinjector.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+          return m_service;
+      }
+
+      void setService(I* service) {
+          m_service = service;
+      }
+
+private:
+      I* m_service;
+};
```
Comment (Jojo-Schmitz) [inline global/serviceinjector.h]: semicolon?

Comment (Jojo-Schmitz) [reply global/serviceinjector.h]: again indent style...

**Comment by Howard-C** [comment]
Also, please change the e-mail address of your git client to the same of you GitHub account, so that these commits can be shown done by this account: @vpereverzev, so that you can have a "Contributor" label beside your descriptions and comments, so that we won't have to ask you to sign the CLA or other troubles every time.

**Comment by Howard-C** [comment]
You still have to get rid of these merge commits. I know this is not very easy but they just don't belong in here.

For further information, you don't have to perform merges at all after you start working on a branch. If you want to update the branch with new commits on master, use `git rebase master` on your branch, it's very efficient and not troublesome. If there are conflicts, resolve them in the conflicted files and use `git add <resolved-files>`. Then the conflicted commit will update itself, and there's no merge commit.

File: mscore/modulessetup.cpp [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    m_modulesSetupList << new TelemetrySetup();
+#endif
+}
+
+void ModulesSetup::setup()
+{
+    for (AbstractModuleSetup* moduleSetup : m_modulesSetupList) {
+         moduleSetup->setup();
+    }
+}
```
Comment (Jojo-Schmitz) [inline mscore/modulessetup.cpp]: Banner style and 6 space indent...

File: mscore/modulessetup.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    void setup();
+
+private:
+    Q_DISABLE_COPY(ModulesSetup)
+
+    ModulesSetup();
+
+    QList<AbstractModuleSetup*> m_modulesSetupList;
+};
```
Comment (Jojo-Schmitz) [inline mscore/modulessetup.h]: Banner style and 6 space indent...

File: thirdparty/google_analytics/ganalytics.h [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    void sendIntervalChanged();
+    void isSendingChanged(bool isSending);
+
+private:
+    class Private;
+    Private *d;
+
+    friend QDataStream& operator<<(QDataStream &outStream, const GAnalytics &analytics);
+    friend QDataStream& operator>>(QDataStream &inStream, GAnalytics &analytics);
+};
```
Comment (Jojo-Schmitz) [inline thirdparty/google_analytics/ganalytics.h]: Banner style and 6 space indent...

Comment (vpereverzev) [reply thirdparty/google_analytics/ganalytics.h]: As you can see it's a 3-dparty code. So i prefer to not modify it.

**Comment by vpereverzev** [comment]
@Howard-C @Jojo-Schmitz  Yes, I realized too late that I had to work through rebase. Any way, thank you for such a detailed comment. Since I came to the project, I am more and more amazed at the responsiveness of the community. You guys are awesome!

**Comment by Howard-C** [comment]
Thank you. The responsive community was one of the main reasons I joined it without hesitance.

File: thirdparty/google_analytics/ganalytics.h (hunk 11 above)
Comment (Jojo-Schmitz) [reply thirdparty/google_analytics/ganalytics.h]: Oh, yes, sorry, I missed this

**Comment by Howard-C** [comment]
Now there're conflicts to resolve ;-) A good time to try `git rebase master`.

File: mscore/preferences.cpp [hunk 12]
```diff
@@ -98,6 +98,8 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_STARTUP_STARTSCORE,                          new StringPreference(":/data/My_First_Score.mscx", false)},
             {PREF_UI_APP_STARTUP_SHOWTOURS,                        new BoolPreference(true)},
             {PREF_APP_WORKSPACE,                                   new StringPreference("Basic", false)},
+            {PREF_APP_TELEMETRY_ALLOWED,                           new BoolPreference(false)},
+            {PREF_APP_STARTUP_TELEMETRY_ACCESS_REQUESTED,          new StringPreference("", false)},
```
Comment (Jojo-Schmitz) [inline mscore/preferences.cpp]: IMHO this one here should be visible in the Advanced settings, like it needed for testing

File: mscore/preferences.cpp [hunk 13]
```diff
@@ -98,6 +98,8 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_STARTUP_STARTSCORE,                          new StringPreference(":/data/My_First_Score.mscx", false)},
             {PREF_UI_APP_STARTUP_SHOWTOURS,                        new BoolPreference(true)},
             {PREF_APP_WORKSPACE,                                   new StringPreference("Basic", false)},
+            {PREF_APP_TELEMETRY_ALLOWED,                           new BoolPreference(false)},
```
Comment (Jojo-Schmitz) [inline mscore/preferences.cpp]: while this here is needed in the Advanced settings only as long as no better place has been found in the main preferences tabs

**Pull request merged**